### PR TITLE
Support for max_contribution_per_user contribution bounding as per #263

### LINKED ISSUE #263: Support max_contribution_per_user contribution bounding
# Context

## Definitions
_Contribution bounding_ is a process of limiting contributions by a single individual (or an entity represented by a privacy key) to the output dataset or its partition. This is key for DP algorithms, since protecting unbounded contributions would require adding infinite noise.

_Cross-partition contribution bounding_ is a procedure which ensures that each individual contributes to a limited number of partitions.

_Per-partition contribution bounding_ is a procedure which ensures that each individual’s contribution to any single partition is bounded.

More details on the terminology is [here](https://pipelinedp.io/key-definitions/).

## How contribution bounding is performed
[DPEngine.aggregate](https://github.com/OpenMined/PipelineDP/blob/41b70a3c7e19b82024e2d0f44842aaab570440bd/pipeline_dp/dp_engine.py#L52) performs DP aggregations (sum, count, mean etc). In order to apply [Laplace/Gaussain mechanisms](https://en.wikipedia.org/wiki/Additive_noise_mechanisms) queries need to have a bounded sensitivity. For this purposes contribution bounding of each individual is required. 

Now contribution bounding is controlled with `max_partitions_contributed` (cross partition) and `max_contributions_per_partition` (per partition) fields of  [AggregateParams](https://github.com/OpenMined/PipelineDP/blob/41b70a3c7e19b82024e2d0f44842aaab570440bd/pipeline_dp/aggregate_params.py#L55).

[Code](https://github.com/OpenMined/PipelineDP/blob/41b70a3c7e19b82024e2d0f44842aaab570440bd/pipeline_dp/dp_engine.py#L253) that performs contribution bounding.

## Computing sensitivities.
For the Laplace mechanism [L1](https://en.wikipedia.org/wiki/Lp_space) sensitivity is required, for the Gaussian mechanism L2 sensitivity is required. `max_contributions` is L1 sensitivity. For integer valued vector `|x|`, it's known that 

`|x|_2 <= |x|_1` (i.e. L2-norm <= L1-norm). 

i.e. we can estimate L2 sensitivity with L1 sensitivity

# Goals
Implement support of **max_contributions** contribution bounding, which consists

1. To introduce parameter `max_contributions` in [AggregateParams](https://github.com/OpenMined/PipelineDP/blob/41b70a3c7e19b82024e2d0f44842aaab570440bd/pipeline_dp/aggregate_params.py#L55).
2. Implement code which ensures that no more than `max_contributions` rows per privacy_id are preserved (probably [_bound_contributions()](https://github.com/OpenMined/PipelineDP/blob/41b70a3c7e19b82024e2d0f44842aaab570440bd/pipeline_dp/dp_engine.py#L253) is a good starting point).
3. Implement computations of sensitivies, namely we need to update finding sensitivities in [_add_random_noise](https://github.com/OpenMined/PipelineDP/blob/bb4046265e155f45cdd7b4ee91b8d595c1089c72/pipeline_dp/dp_computations.py#L150) based on section computing sensitivities

### PR DESCRIPTION
**Pull request by manavgarg**
## Description
This PR introduces support for a max contributions per user bounding. This PR fixes #263.

## Affected Dependencies
N/A

## How has this been tested?
Test cases have been added for the newly added method.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: pipeline_dp/aggregate_params.py [hunk 1]
```diff
@@ -72,6 +74,7 @@ class AggregateParams:
   """
 
     metrics: Iterable[Metrics]
+    max_contributions: int
```
[the remark this answers is not in the document]
Comment (manavgarg) [reply pipeline_dp/aggregate_params.py:77]: This make sense. I did not know that these two types on contributions are mutually exclusive. Hence, I was also wondering on the relative ordering, whether we should first do per partition bounding followed by total contributions or vice versa. But this explains, thanks.
Will make the changes soon.

File: pipeline_dp/dp_engine.py [hunk 2]
```diff
@@ -269,7 +269,7 @@ def _add_empty_public_partitions(self, col, public_partitions,
 
     def _bound_contributions(self, col, max_partitions_contributed: int,
                              max_contributions_per_partition: int,
-                             aggregator_fn):
+                             max_contributions: int, aggregator_fn):
```
[the remark this answers is not in the document]
Comment (manavgarg) [reply pipeline_dp/dp_engine.py:272]: Agreed, will split it out in a separate method.

File: pipeline_dp/aggregate_params.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (manavgarg) [reply pipeline_dp/aggregate_params.py]: Made the changes.

File: pipeline_dp/dp_engine.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (manavgarg) [reply pipeline_dp/dp_engine.py]: Made the changes.

File: pipeline_dp/dp_engine.py [hunk 3]
```diff
@@ -288,6 +292,37 @@ def _add_empty_public_partitions(self, col, public_partitions,
...
+            partition_key, value).
+          max_contributions: maximum number of records that one privacy id can
+            contribute.
+          aggregator_fn: function that takes a list of values and returns an
+            aggregator object which handles all aggregation logic.
+
+        return: collection with elements ((privacy_id, partition_key),
+              accumulator).
+        """
+        # Max contributions bounding
```
Comment (dvadym) [inline pipeline_dp/dp_engine.py:310]: Nit: Max **user** contribution bounding.

File: pipeline_dp/dp_engine.py [hunk 4]
```diff
@@ -288,6 +292,37 @@ def _add_empty_public_partitions(self, col, public_partitions,
...
+        col = self._backend.sample_fixed_per_key(col, max_contributions,
+                                                 "Sample per privacy_id")
```
Comment (dvadym) [inline pipeline_dp/dp_engine.py:312]: Nit: add
\# (privacy_id, [partition_key, value])

File: pipeline_dp/dp_engine.py [hunk 5]
```diff
@@ -288,6 +292,37 @@ def _add_empty_public_partitions(self, col, public_partitions,
...
+        self._add_report_stage(
+            f"max contribution bounding: randomly selected not "
+            f"more than {max_contributions} contributions")
+        col = self._backend.map_tuple(
```
Comment (dvadym) [inline pipeline_dp/dp_engine.py:316]: In order to be consistent with `_contribution_bounding` we shoud aggregate per (privacy_id, partition), nicely that we already grouped per `privacy_id` (and sumsampled), we can assumet that `max_contributions` is small enough to make everything in memory for one `(privacy_id, )`. 
Namely here we can use `map_values` with `fn` which takes a list ` [(partition_key, value)]` and this `fn` groups by `partition_key` (e.g. with dictionaries) and then applies `aggregator_fn` for values per each.

Does it make sense?

File: pipeline_dp/aggregate_params.py [hunk 6]
```diff
@@ -141,6 +144,13 @@ def __post_init__(self):
             raise ValueError(
                 "AggregateParams.public_partitions is deprecated. Please use public_partitions argument in DPEngine.aggregate insead."
             )
+        if self.max_contributions is not None:
+            if ((self.max_partitions_contributed is not None) or
+                (self.max_contributions_per_partition is not None)):
+                raise ValueError(
+                    "Only one in params.max_contributions or "
+                    "(params.max_partitions_contributed and "
+                    "params.max_contributions_per_partition) must be set")
```
[the remark this answers is not in the document]
Comment (manavgarg) [reply pipeline_dp/aggregate_params.py:153]: Do you want to check for the above when no max_contributions is specified?
Are we enforcing the user to provide atleast one kind of contribution binding?

File: pipeline_dp/aggregate_params.py [hunk 7]
```diff
@@ -141,6 +144,13 @@ def __post_init__(self):
             raise ValueError(
                 "AggregateParams.public_partitions is deprecated. Please use public_partitions argument in DPEngine.aggregate insead."
             )
+        if self.max_contributions is not None:
+            if ((self.max_partitions_contributed is not None) or
```
[the remark this answers is not in the document]
Comment (manavgarg) [reply pipeline_dp/aggregate_params.py:148]: Done.

File: pipeline_dp/dp_engine.py (hunk 3 above)
Comment (manavgarg) [reply pipeline_dp/dp_engine.py]: Done.

File: pipeline_dp/dp_engine.py (hunk 4 above)
Comment (manavgarg) [reply pipeline_dp/dp_engine.py:312]: Done.

File: pipeline_dp/dp_engine.py (hunk 5 above)
Comment (manavgarg) [reply pipeline_dp/dp_engine.py:317]: I don't think I fully understand this. Possible to discuss this sometime?

File: pipeline_dp/aggregate_params.py [hunk 8]
```diff
@@ -77,8 +79,9 @@ class AggregateParams:
   """
 
     metrics: Iterable[Metrics]
-    max_partitions_contributed: int
-    max_contributions_per_partition: int
+    max_contributions: Optional[int]
```
Comment (dvadym) [inline pipeline_dp/aggregate_params.py:82]: Please add `= None` all 3 fields

File: pipeline_dp/aggregate_params.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dvadym) [reply pipeline_dp/aggregate_params.py:153]: Correct. Previously contributiong bounding fields were non-optional, so there was no need to check

File: pipeline_dp/aggregate_params.py (hunk 8 above)
Comment (manavgarg) [reply pipeline_dp/aggregate_params.py]: Done.

File: pipeline_dp/dp_engine.py (hunk 5 above)
Comment (manavgarg) [reply pipeline_dp/dp_engine.py:347]: Thanks for the pointers. Have fixed the implementation and have also added test cases to verify this.

File: pipeline_dp/aggregate_params.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (manavgarg) [reply pipeline_dp/aggregate_params.py:153]: Thanks. Done.

File: pipeline_dp/dp_engine.py [hunk 9]
```diff
@@ -288,6 +293,71 @@ def _add_empty_public_partitions(self, col, public_partitions,
...
+        # and value to be list of input values.
+        def collect_values_per_partition_key_per_privacy_id(input_list):
+            d = collections.defaultdict(list)
+            for key, value in input_list:
+                d[key].append(value)
+            return d
+
+        col = self._backend.map_values(
+            col, collect_values_per_partition_key_per_privacy_id,
+            "Apply fn to convert per privacy id values into a dict of values"
```
Comment (dvadym) [inline pipeline_dp/dp_engine.py:333]: Nit: "Group per (privacy_id, partition_key)." 

It's better to keep stage name shorter, it's more convenient to check them in moninoring UI

**Review by dvadym — COMMENTED** [review]
Thanks a lot, cointribution bounding looks good! 

I left some minor comments. And after addressing those comments, this PR can be landed.Let's do the last part (of computing sensitivities) in the subsequent PR.

File: pipeline_dp/dp_engine.py [hunk 10]
```diff
@@ -288,6 +293,71 @@ def _add_empty_public_partitions(self, col, public_partitions,
...
+        # Rekey it into values per privacy id and partition key. This returns
+        # it as one list per privacy id.
+        def rekey_per_privacy_id_per_partition_key(privacy_id, partition_dict):
+            result = list()
+            for partition_key in partition_dict:
+                result.append(((privacy_id, partition_key),
+                               partition_dict[partition_key]))
+            return result
+
+        col = self._backend.map_tuple(
```
[the remark this answers is not in the document]
Comment (manavgarg) [reply pipeline_dp/dp_engine.py:343]: I am confused here. I think we need to do map_tuple first to convert first group it by privacy_id, partition_key ? Not sure how I can apply flat_map before that.

File: pipeline_dp/dp_engine.py (hunk 9 above)
Comment (manavgarg) [reply pipeline_dp/dp_engine.py]: Done.

File: pipeline_dp/dp_engine.py [hunk 11]
```diff
@@ -288,6 +293,71 @@ def _add_empty_public_partitions(self, col, public_partitions,
...
+            "Apply fn to convert per privacy id values into a dict of values"
+            "per partition key")
+
+        # (privacy_id, {partition_key: [value]})
+
+        # Rekey it into values per privacy id and partition key. This returns
+        # it as one list per privacy id.
+        def rekey_per_privacy_id_per_partition_key(privacy_id, partition_dict):
+            result = list()
+            for partition_key in partition_dict:
```
[the remark this answers is not in the document]
Comment (manavgarg) [reply pipeline_dp/dp_engine.py]: Done.

File: pipeline_dp/dp_engine.py [hunk 12]
```diff
@@ -104,10 +105,14 @@ def _aggregate(self, col, params: pipeline_dp.AggregateParams,
                                   data_extractors.value_extractor(row)),
                 "Extract (privacy_id, partition_key, value))")
             # col : (privacy_id, partition_key, value)
-            col = self._bound_contributions(
-                col, params.max_partitions_contributed,
-                params.max_contributions_per_partition,
-                combiner.create_accumulator)
+            if params.max_contributions is not None:
```
[the remark this answers is not in the document]
Comment (manavgarg) [reply pipeline_dp/dp_engine.py]: Done.

File: pipeline_dp/dp_engine.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dvadym) [reply pipeline_dp/dp_engine.py:343]: I meant, that it's possible to combine map_tuple+flat_map in one `flat_map`, namely

```
  def rekey_per_privacy_id_per_partition_key(pid_pk_dict):
          privacy_id, partition_dict = pid_pk_dict
          for partition_key, values in partition_dict.items():
              yield (privacy_id, partition_key), values

   # Unnest the list per privacy id.
   col = self._backend.flat_map(col, rekey_per_privacy_id_per_partition_key,"Unnest")
```

[the remark this answers is not in the document]
Comment (manavgarg) [reply pipeline_dp/dp_engine.py]: Thanks for the clarification and the pointers. Have made the change.

**Review by dvadym — APPROVED** [review]
Thank you!

**Pull request merged**